Add new configuration `NumberOfEmptyLines` for `Style/EmptyLineBetweenDefs`

Dorian opened the pull request:
Add new configuration `NumberOfEmptyLines` for `Style/EmptyLineBetweenDefs` 

Allows config for multiple consecutive empty lines, e.g. `NumberOfEmptyLines: 2`.

```ruby
# bad
def a
end

def b
end

# good
def a
end


def b
end
```

Also supports minimum and maximum (e.g. `NumberOfEmptyLines: [1, 2]`)

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [ ] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

Dorian commented:
Also, @bbatsov, are the extra trailing spaces in `manual/cops_style.md` (e.g. line 18, 217, 465, etc.) on purpose? It doesn't seem to make a difference: https://gist.github.com/Dorian/68c20fd65f1520035e92aa8a778df203

(e.g. my vim config removes extra trailing spaces by default)

mikegee commented:
I'm surprised this is something you want. Could you please explain it more?

The style guide might contradict this a bit: https://github.com/bbatsov/ruby-style-guide#two-or-more-empty-lines

Dorian commented:
@mikegee Yes, I don't personally agree with it, but that's just the way it is. Configurations options are exactly there for people who deviate from the normal style guide in my opinion, so this seems legit.

bbatsov commented:
I'm on the fence about this, as I don't like adding options that'd be useful to just a handful of people. The increased complexity of the code is simply not worth it. At any rate `MinimumEmptyLines` is a bad name for this - it should be `RequiredEmptyLines` or something like this. Allowing a flexible number of empty lines between method definitions seems like a really idea to me for the code consistency.

Dorian commented:
@bbatsov The previous behavior was to require a minimum of 1 empty line between, e.g. this test pass:

```ruby
  it 'allows more than the required number of newlines' do
    source = ['    def n',
              '    end',
              '    ',
              '    ',
              '    def o',
              '    end']
    inspect_source(cop, source)
    expect(cop.offenses).to be_empty
  end
```

Maybe having a minimum and a maximum would be better, e.g. people might want 1 or 2 lines but no more, or 0 or 1 line, or 0 only, or 1 only or 2 only, etc. I'm gonna implement it this way because it also improves the 1 line case

Dorian commented:
@bbatsov @mikegee What do you think now? It even autocorrects down to the number of newlines required :)

(e.g. worst case scenario for this PR, I make the `Style/EmptyLineBetweenDefs` better)

bbatsov commented:
> @bbatsov The previous behavior was to require a minimum of 1 empty line between, e.g. this test pass:

Hmm, I guess I didn't notice it. In light of this your current suggestion makes sense. Even though I don't like using varying number of blank lines, probably due to this oversight some people were using them, so your changes look like a step in the right direction.

bbatsov commented:
You should extend the class documentation a bit.

mikegee commented:
This PR is implemented well, and fixing the existing bug is nice. I still think it is weird that nobody has explicitly requested this behavior.

bbatsov commented:
> This PR is implemented well, and fixing the existing bug is nice. I still think it is weird that nobody has explicitly requested this behavior.

In the absence of the bug someone might have asked for it. :-)

Dorian commented:
@bbatsov @mikegee Ok, updated docs, docs generation doesn't let me make double empty lines, not sure what to do

bbatsov commented:
> @bbatsov @mikegee Ok, updated docs, docs generation doesn't let me make double empty lines, not sure what to do

I'm not sure what do you mean by this.

Dorian commented:
In the class documentation I put two newlines as an example but in the
generated docs it only has one empty line, will fix in my other markdown pr
On Thu, 2 Mar 2017 at 21:50, Bozhidar Batsov <notifications@github.com>
wrote:

> @bbatsov <https://github.com/bbatsov> @mikegee
> <https://github.com/mikegee> Ok, updated docs, docs generation doesn't
> let me make double empty lines, not sure what to do
>
> I'm not sure what do you mean by this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bbatsov/rubocop/pull/4083#issuecomment-283871808>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAH4faSZiUqGt6tCXTRAphPIKIgRpN-fks5rh6oqgaJpZM4MQEJr>
> .
>

Dorian commented:
@bbatsov Anyway, looking good now :)

Dorian commented:
(got to handle the case of non def lines in between (E.g. private / comments / any code?)

Dorian commented:
So, still need to fix auto-correct when there are comments in the middle, gonna leave the "`private`/`public` + comments" case to `EmptyLinesAroundAccessModifier` (that I need to work on too after that).

```
def a
end
private
# hello
def b
end
```

Dorian commented:
Ok, it just ignores the the groups edge cases, e.g.

```ruby
def a
end

# comment

def b
end
```

bbatsov commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -247,6 +258,17 @@
     expect(cop.offenses).to be_empty
   end
 
+  it 'allows more than the required number of newlines' do
```
doesn't allow

bbatsov commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -268,4 +290,114 @@
   end
+
+  context 'when a maximum of empty lines is specified' do
+    let(:cop_config) { { 'MinimumEmptyLines' => 0, 'MaximumEmptyLines' => 1 } }
```
I do wonder if we should allow 0 as a valid configuration value, but I guess I can live with this.

bbatsov commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -235,6 +235,17 @@
   end
 
+  it 'auto-corrects when there are too many new lines' do
+    corrected = autocorrect_source(cop, ['  def a; end',
```
Why the leading whitespace?

bbatsov commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -247,6 +258,17 @@
   end
 
+  it 'allows more than the required number of newlines' do
+    source = ['    def n',
```
Here as well.

bbatsov commented on spec/rubocop/cop/style/empty_line_between_defs_spec.rb:
```diff
@@ -268,4 +290,114 @@
+    let(:cop_config) { { 'MinimumEmptyLines' => 0, 'MaximumEmptyLines' => 1 } }
+
+    it 'finds no offense for no empty line' do
+      source = ['    def n',
```
And here.

deivid-rodriguez commented:
Why do we want Min and Max config? Wouldn't just a single (exact) number of lines setting suffice? In other words, is enforcing a bounded but inconsistent number of empty lines something anyone would want?

bbatsov commented:
I think some people have a tendency to separate groups of related methods with 2 lines, but I'm not sure what @Dorian's particular coding style is.

deivid-rodriguez commented:
Yeah, maybe...

My concern (other that the need for the feature which I'm not sure either) is that adding this makes the feature harder to understand (because of naming). It's not obvious that min will default to max if max is not present and one might think that things like

```
Style/EmptyLineBetweenDefs
  MinimumEmptyLines: 2
```

require further tweaking to get a max configuration.

I think I'd by -1 for the min/max feature, but if we're adding it, what about a single config which can be a single number or an array?

```
Style/EmptyLineBetweenDefs
  NumberOfEmptyLines: 1
```
```
Style/EmptyLineBetweenDefs
  NumberOfEmptyLines: [1, 2]
```

bbatsov commented:
I think we've got plenty of cops with confusing configuration. :D But I see your point. I was also reminded the name of the cop should probably be updated as well - originally it was simply enforcing a single empty line between defs (at least it was supposed to, turned out it was buggy).

bbatsov commented:
@Dorian you changed the code, but not the commit message.

Dorian commented:
> I think some people have a tendency to separate groups of related methods with 2 lines, but I'm not sure what @Dorian's particular coding style is.

@deivid-rodriguez @bbatsov It's the project's coding style to separate methods with two newlines, I also see people separating sections of related methods with two new lines, e.g.:

```ruby
def handle_apples
end

def handle_pears
end


def process_apples
end

def process_pears
end
```

---

@deivid-rodriguez I actually like your `NumberOfEmptyLines` better, so I changed it to that

---

> the name of the cop should probably be updated as well

@bbatsov The default being one line, I'm thinking of keeping it `Style/EmptyLineBetweenDefs`.

---

@bbatsov  About the extra empty spaces in source examples, I removed them when there were not specified by the spec description.
